### PR TITLE
fix(picotron): address reproducibility, documentation, and EFA reliability (#975)

### LINKED ISSUE #975: [Feature]: Harden picotron test case — pin version, fix docs, add EFA/NCCL config
### Feature Description & Motivation

The picotron test case (`3.test_cases/pytorch/picotron/`) has several issues that make builds non-reproducible, documentation inaccurate, and multi-node runs unreliable on EFA-equipped instances. Other test cases in the repo (e.g., FSDP, DeepSpeed, TRL) have matured past these problems and can serve as reference implementations.

**Key problems:**

1. **Non-reproducible Docker builds** — the Dockerfile clones `huggingface/picotron` at HEAD (`git clone https://github.com/huggingface/picotron`) with no pinned tag, branch, or commit hash. Upstream picotron has no releases or tags, so any upstream change silently alters the built image. Upstream is also pinned to `torch==2.1.0` (Oct 2023), which is considerably older than what other test cases use.

2. **Documentation path mismatch** — the Slurm README (`SmolLM-1.7B/slurm/README.md`) says the config will be created at `./conf/llama-1B-dp2-tp2-pp2/config.json`, but the `create_config.py` command shown uses `--exp_name llama-1B-tp2`, which produces `./conf/llama-1B-tp2/config.json`. The sbatch script correctly references `llama-1B-tp2`, so the README text is wrong.

3. **Duplicate prerequisite in EC2 README** — `SmolLM-1.7B/ec2/README.md` lists "Build the `picotron` container image" as both prerequisite #1 and prerequisite #3.

4. **No EFA/NCCL tuning for multi-node** — `train.sbatch` has no EFA environment variables (`FI_PROVIDER`, `FI_EFA_USE_HUGE_PAGE`, etc.) and no NCCL tuning (`NCCL_DEBUG`, `NCCL_BUFFSIZE`, `NCCL_TUNER_PLUGIN`). Other test cases (FSDP, TRL) set these. Without them, multi-node runs on EFA-equipped instances (p4d, p5) may fall back to TCP or hit silent performance issues.

5. **`create_config.py` depends on upstream's `template/base_config.json`** — this file is not in the ADT repo; it comes from the `git clone` inside the container. This is technically correct (the Dockerfile sets `WORKDIR /picotron`) but is confusing and means `create_config.py` cannot be used outside the container without also cloning picotron.

### Category

Enhancement to existing test case

### Alternatives Considered

_No alternatives — these are improvements to the existing test case._

#### Affected files

| File | Issue |
|------|-------|
| `3.test_cases/pytorch/picotron/picotron.Dockerfile` | Unpinned `git clone` of picotron; `torch==2.1.0` |
| `3.test_cases/pytorch/picotron/SmolLM-1.7B/slurm/README.md` | Config path says `llama-1B-dp2-tp2-pp2` but should be `llama-1B-tp2` |
| `3.test_cases/pytorch/picotron/SmolLM-1.7B/ec2/README.md` | Duplicate prerequisite (#1 and #3 both say "build the container") |
| `3.test_cases/pytorch/picotron/SmolLM-1.7B/slurm/train.sbatch` | Missing EFA/NCCL environment variables for multi-node |
| `3.test_cases/pytorch/picotron/create_config.py` | Implicit dependency on upstream `template/base_config.json` |

#### Upstream reference

- Upstream repo: [huggingface/picotron](https://github.com/huggingface/picotron) — no releases/tags; latest commit `59714b1bf255` (2025-08-26)
- Upstream pins `torch==2.1.0`, `triton==2.1.0`, `flash-attn==2.5.0`, `transformers==4.47.0`
- Upstream has no Dockerfile — the one in ADT is the only containerized build

#### Suggested fixes

| # | Fix | Effort |
|---|-----|--------|
| 1 | Pin picotron to a specific commit hash in the Dockerfile (e.g., `git clone ... && cd picotron && git checkout 59714b1bf255`) | Small |
| 2 | Fix the config path in the Slurm README | Trivial |
| 3 | Remove the duplicate prerequisite in the EC2 README | Trivial |
| 4 | Add standard EFA/NCCL environment variables to `train.sbatch` (see FSDP's sbatch files for reference) | Small |
| 5 | Either vendor `template/base_config.json` into the ADT repo or document the container-only constraint in the README | Small |

#### Patterns from other test cases

For reference, the FSDP and DeepSpeed test cases:
- Pin all upstream repos to specific versions/tags/commits
- Set EFA environment variables (`FI_PROVIDER=efa`, `FI_EFA_USE_HUGE_PAGE=0`, `NCCL_DEBUG=WARN`)
- Separate `--output` and `--error` in sbatch
- Provide per-model config generation with Jinja2 templates

### PR DESCRIPTION
**Pull request by littlemex**
## Purpose

Fixes #975: five issues affecting reproducibility, documentation accuracy, and multi-node reliability of the picotron test case.

## Changes

- **`picotron.Dockerfile`**: Pin picotron clone to commit `59714b1bf255` to prevent silent upstream changes from altering the built image.

  Note: `torch==2.1.0` is intentionally left unchanged. The pinned commit's `requirements.txt` specifies `torch==2.1.0`, so a version bump in the Dockerfile alone would be silently overridden by `pip install -e .`.

- **`SmolLM-1.7B/slurm/README.md`**: Fix documented config output path (`llama-1B-dp2-tp2-pp2` → `llama-1B-tp2`) to match the actual `--exp_name` argument and the path used in `train.sbatch`.

- **`SmolLM-1.7B/ec2/README.md`**: Remove duplicate "Build the container image" prerequisite (was listed as both item 1 and item 3).

- **`SmolLM-1.7B/slurm/train.sbatch`**: Add EFA/NCCL environment variables for reliable multi-node performance. Variable selection follows [`1.architectures/efa-cheatsheet.md`](https://github.com/awslabs/awsome-distributed-training/blob/main/1.architectures/efa-cheatsheet.md).

  | Variable | Reason |
  |----------|--------|
  | `FI_EFA_USE_HUGE_PAGE=0` | Preset 3.1 (libfabric>=1.18.0, aws-ofi-nccl>=1.7.0). Prevents `"Cannot allocate memory"` from PyTorch multi-process dataloader exhausting huge pages. |
  | `FI_PROVIDER=efa` | Explicitly selects the EFA provider so the script does not depend on a specific host aws-ofi-nccl version to function correctly. |
  | `NCCL_SOCKET_IFNAME=en` | Restricts NCCL socket rendezvous to ENA/EFA interfaces (`ens*`, `en*`), avoiding accidental use of `docker0`/`lo`/`veth`. |
  | `NCCL_DEBUG=INFO/WARN` (omitted) | Cheatsheet: "we advise turning it off unless you suspect NCCL issues." Maintainers may add `NCCL_DEBUG=WARN` if this script is used for multi-node troubleshooting. |

- **`create_config.py` + `template/base_config.json`**: `create_config.py` opened `template/base_config.json` relative to CWD, so it only worked inside the container. Fix resolves the path relative to `__file__` and vendors the template into this repository (sourced from the pinned commit `59714b1bf255`), so the script works both inside and outside the container.

## Test Plan

**Environment:**
- AWS Service: AWS ParallelCluster 3.13.1 / Slurm 24.05.8
- Instance type: g5.8xlarge (NVIDIA A10G, EFAv2)
- Number of nodes: 2

Note: Validation was performed on `g5.8xlarge` rather than the primary target instances (`p4d`/`p5`) due to resource availability and cost constraints. The fixes themselves are not instance-specific; EFA transport, NCCL behaviour, and the Python path fix apply equally across instance families.

**Test commands:**
```bash
# Build container image
cd /fsx/picotron
docker build -t picotron -f picotron.Dockerfile .

# Convert to enroot squash image for Slurm
enroot import -o /fsx/picotron/picotron.sqsh dockerd://picotron:latest

# Generate training config from outside the container (verifies issue #5)
enroot create --name picotron /fsx/picotron/picotron.sqsh
enroot start \
  --mount /fsx/picotron/SmolLM-1.7B/slurm:/fsx/picotron/SmolLM-1.7B/slurm \
  --env NVIDIA_VISIBLE_DEVICES=void picotron \
  python3 /picotron/create_config.py \
    --out_dir /fsx/picotron/SmolLM-1.7B/slurm/conf \
    --exp_name llama-1B-tp2 --dp 1 --tp 2 --pp 1 \
    --model_name HuggingFaceTB/SmolLM-1.7B \
    --num_hidden_layers 5 --total_train_steps 5000 \
    --grad_acc_steps 2 --mbs 4 --seq_len 128
enroot remove -f picotron

# Submit training job
cd /fsx/picotron/SmolLM-1.7B/slurm
mkdir -p logs && ln -sf /fsx/picotron/picotron.sqsh picotron.sqsh
sbatch train.sbatch
```

## Test Results

**Dockerfile commit pin (issue #1)**

```
HEAD is now at 59714b1 Merge pull request #31 from phi9t/patch-1
```

**Config path (issue #2)**

`conf/llama-1B-tp2/config.json` created as expected.

**EFA provider selection (issue #4) — `NCCL_DEBUG=INFO` on 2-node torchrun all-reduce**

```
NCCL INFO NET/OFI Initializing aws-ofi-nccl 1.13.2-aws
NCCL INFO NET/OFI Selected Provider is efa (found 1 nics)
NCCL INFO NET/Libfabric : GPU Direct RDMA Disabled for HCA 0 'rdmap0s29'
NCCL INFO Channel 00/0 : 0[0] -> 1[0] [send] via NET/Libfabric/0
[PASS] all_reduce complete: world_size=2, result=2.0
```

**2-node TP=2 training job — `sbatch train.sbatch` (5000 steps, no NCCL errors)**

```
0: [rank 0] Step:    1 | Loss: 10.9688 | Tokens/s:   215.85 | Memory: 2.75GB
0: [rank 0] Step:  100 | Loss:  4.2344 | Tokens/s:  7297.38 | Memory: 2.85GB
0: [rank 0] Step:  500 | Loss:  3.3750 | Tokens/s:  7358.21 | Memory: 2.85GB
0: [rank 0] Step: 1000 | Loss:  3.4062 | Tokens/s:  7230.00 | Memory: 2.85GB
0: [rank 0] Step: 5000 | Loss:  2.5547 | Tokens/s:  7250.00 | Memory: 2.85GB
```

Job completed with exit code 0. No NCCL errors throughout all 5000 steps.

## Directory Structure

```
3.test_cases/
└── pytorch/
    └── picotron/
        └── SmolLM-1.7B/
            ├── picotron.Dockerfile
            ├── create_config.py
            ├── template/
            │   └── base_config.json   # new: vendored from picotron@59714b1
            ├── VALIDATION.md          # new: validation procedure and results
            ├── slurm/
            │   ├── README.md
            │   └── train.sbatch
            └── ec2/
                └── README.md
```

## Checklist

- [x] I have read the [contributing guidelines](https://github.com/awslabs/awsome-distributed-training/blob/main/CONTRIBUTING.md).
- [x] I am working against the latest `main` branch.
- [x] I have searched existing open and recently merged PRs to confirm this is not a duplicate.
- [x] The contribution is self-contained with documentation and scripts.
- [x] External dependencies are pinned to a specific version or tag (no `latest`).
- [x] A README is included or updated with prerequisites, instructions, and known issues.

### CODE REVIEW

File: 3.test_cases/pytorch/picotron/SmolLM-1.7B/slurm/train.sbatch [hunk 1]
```diff
@@ -16,6 +16,15 @@ set -ex;
...
 GPUS_PER_NODE=1
 
+###########################
+######## EFA / NCCL #######
+###########################
+
+## See 1.architectures/efa-cheatsheet.md for full reference
+export FI_EFA_USE_HUGE_PAGE=0
+export FI_PROVIDER=efa
+export NCCL_SOCKET_IFNAME=en
```
Comment (KeitaW) [inline 3.test_cases/pytorch/picotron/SmolLM-1.7B/slurm/train.sbatch]: ### `NCCL_SOCKET_IFNAME` — I'd suggest using an exclusion pattern instead of positive selection

The value `en` is a positive-match pattern (selects interfaces starting with `en`). The repository convention — documented in the [EFA cheatsheet](https://github.com/awslabs/awsome-distributed-training/blob/main/1.architectures/efa-cheatsheet.md) and used consistently across FSDP, DeepSpeed, and nanoVLM sbatch scripts — is to use **exclusion-based patterns** instead.

Positive selection can break on instances where the EFA interfaces don't match the `en*` prefix (e.g., `efa0`, `efa1` on p4d/p5 instances). The exclusion pattern is safer because it blocks known-bad interfaces while allowing any legitimate network interface — including EFA devices — to be used.

This matches the pattern used in `3.test_cases/pytorch/FSDP/slurm/*.sbatch`, `3.test_cases/pytorch/deepspeed/gpt/slurm/*.sbatch`, and `3.test_cases/pytorch/nanoVLM/slurm/*.sbatch`.

```suggestion
export NCCL_SOCKET_IFNAME=^docker,lo,veth
```
